examples/Tutorial 4 - 3SAT Solver.ipynb

```diff
@@ -0,0 +1,268 @@
+    "\n",
+    "- Once we have a MIS solution, we only need to confirm that one node in each clause is part of the MIS solution, i.e. we only need to check the length of the MIS solution (Since the maximum number of nodes is the number of clauses, and that in each triangle representing a clause we can have at most one node, reaching a configuration with exactly the good number of nodes implies that the solution is valid).\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 85,
+   "id": "99bce83e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class ThreeSATTerm:\n",
+    "    \"\"\"\n",
+    "    A 3SAT term, i.e. a variable (for instance \"x3\") and possibly a negation.\n",
+    "    \"\"\"\n",
+    "    variable: str\n",
+    "    positive: bool\n",
+    "    \n",
+    "    def __init__(self, variable, positive):\n",
+    "        self.variable = variable\n",
+    "        self.positive = positive\n",
+    "    \n",
+    "    def eval(self, values: dict[str, bool]) -> bool:\n",
+    "        \"\"\"\n",
+    "        Evaluate this term with the value of the variables as specified in `values`.\n",
+    "        \"\"\"\n",
+    "        if self.variable not in values:\n",
+    "            raise ValueError(f\"Variable '{self.variable}' not found in values.\")\n",
+    "        \n",
+    "        return values[self.variable]\n",
+    "    \n",
+    "    \n",
+    "\n",
+    "class ThreeSATClause:\n",
+    "    \"\"\"\n",
+    "    A 3SAT clause, i.e. T1 OR T2 OR T3 where T1, T2 and T3 are exactly three 3SAT terms.\n",
+    "    \"\"\"\n",
+    "\n",
+    "    terms: list[ThreeSATTerm]\n",
+    "    \n",
+    "    def __init__(self, terms):\n",
+    "        if len(terms) != 3:\n",
+    "            raise ValueError(\"A 3SAT clause must have exactly three terms.\")\n",
+    "        self.terms = terms\n",
+    "\n",
+    "    def eval(self, values: dict[str, bool]) -> bool:\n",
+    "        \"\"\"\n",
+    "        Evaluate this clause with the value of the variables as specified in `values`.\n",
+    "        Returns `True` if and only for any `term`, `term.eval(values)` returns `True`.\n",
+    "        \"\"\"\n",
+    "        if len(self.terms) != 3:\n",
+    "            raise ValueError(\"A 3SAT clause must have exactly three terms.\")\n",
+    "        for term in self.terms:\n",
+    "            if term.eval(values):\n",
+    "                return True\n",
+    "        return False\n",
+    "        pass\n",
+    "    \n",
+    "    \n",
+    "\n",
+    "\n",
+    "class ThreeSATInstance:\n",
+    "    \"\"\"\n",
+    "    A 3SAT instance, i.e. C1 AND C2 AND C3  AND Cn where each Ci is a 3SAT Clause.\n",
+    "    \"\"\"\n",
+    "\n",
+    "    clauses: list[ThreeSATClause]\n",
+    "    _variables: list[str]\n",
+    "    \n",
+    "    def __init__(self, clauses):\n",
+    "        self.clauses = clauses\n",
+    "        self._variables = []\n",
```

Comment:
The same variable can appear in various clauses, so this should probably be a `set`.